File: src/main/java/de/flapdoodle/generics/local/j/HasLocalTypeMethod.java
package de.flapdoodle.generics.local.j;

import java.util.function.Supplier;

public class HasLocalTypeMethod {

  public <T> T outer(Factory<T,?> factory, T value) {
    return sample(factory, value);
  }

  public <T,R extends Supplier<T>> T sample(Factory<T,R> factory, T value) {
    R supplier = factory.create(value);
    return supplier.get();
  }
}

File: src/main/kotlin/de/flapdoodle/generics/local/k/HasLocalTypeMethod.kt
package de.flapdoodle.generics.local.k

import java.util.function.Supplier

class HasLocalTypeMethod {

  fun <T> outer(factory: Factory<T, *>, value: T): T {
    @Suppress("UNCHECKED_CAST")
    return sample(factory as Factory<T, Supplier<T>>, value)
  }

  fun <T, R : Supplier<T>> sample(factory: Factory<T, R>, value: T): T {
    val supplier = factory.create(value)
    return supplier.get()
  }
}
